web/packages/teleterm/src/ui/utils/retryWithRelogin.ts

```diff
@@ -75,7 +81,13 @@ export async function retryWithRelogin<T>(
 
   const rootClusterUri = routing.ensureRootClusterUri(resourceUri);
 
-  await login(appContext, rootClusterUri);
+  if (!pendingLogin) {
+    pendingLogin = login(appContext, rootClusterUri).finally(() => {
+      pendingLogin = undefined;
+    });
+  }
```

Comment:
This doesn't handle well two concurrent requests from two different workspaces. Let's handle this scenario and add a test case for it.

At the moment if workspace B calls `retryWithRelogin` while the call from workspace A is in progress, the call from workspace B will wait for the call from A to resolve, after which it'll call `actionToRetry()` again. It'll fail because we didn't attempt to log in to workspace B, we just waited for the workspace A to finish its login ceremony.

I think the way to address this is to change `retryWithRelogin` so that if there's a pending login, it should wait for the login to finish. Then it should check if a relogin is still needed and if so, show the modal again.

There are two problems though:

1. At the moment the only way `retryWithRelogin` knows whether a relogin is needed is when `actionToRetry` returns a retryable error.
   * This is not ideal because we want to know this without retrying an action.
   * We cannot depend on the state of `ClustersService`, because `cluster.connected` does not get flipped to false after the cert expires.
   * Perhaps we could do a quick `tshd.getCluster` to check if the cert is valid? This doesn't make any network requests and returns current `connected` so it should be quick.
2. If there's one login in progress and two `retryWithRelogin` waiting for it to finish, they cannot both just wait for the login to finish and then show the modal at the same time. The second one to run would just override the other one that was waiting.
   * One way to solve it would be to make a recursive call to `retryWithRelogin` if we fetch fresh `connected` and deem that a relogin is still necessary, instead of just calling `actionToRetry`. We could still call `actionToRetry` if it turned out that relogin is not necessary anymore. 
3. If we addressed 1 and 2, if there were two concurrent calls from the same workspace and the user canceled the first one, the second one would show a login modal again. I feel like in this situation it should just fail.
   * I guess on top of storing `pendingLogin`, we could also store which workspace this `pendingLogin` is related to by storing the root cluster URI. Then if we see that the URI is the same, the code can use the current logic from your PR. If the URIs are different, it should trigger the logic with showing another login modal.
   * Mind you that here you'd have the same problem with showing two modals in quick succesion that we had when working with Brian to implement headless login.

Reply:
Thanks for the deep analysis!

The code I added indeed doesn't support the scenario with multiple workspaces, but I thought it wasn't really necessary.

Before we make an attempt to relogin (which shows a modal) in `retryWithRelogin` we do this:
```ts
  if (!workspacesService.doesResourceBelongToActiveWorkspace(resourceUri)) {
    // The error is retryable, but by the time the request has finished, the user has switched to
    // another workspace so they're no longer looking at the relevant UI.
    //
    // Since it might take a few seconds before the execution gets to this point, the user might no
    // longer remember what their intent was, thus showing them a login modal could be disorienting.
    //
    // In that situation, let's just not attempt to relogin and instead let's surface the original
    // error.
    throw retryableErrorFromActionToRetry;
  }
```
Doesn't it solve our problem in the first place?
If you run an action while you are in a workspace A, it will fail without showing a login modal if you have switched to workspace B.

If I'm correct, the only situation in which we would have to deal with handling concurrent requests from different workspaces would be the following scenario:
1. The request from workspace A triggers a login modal, `pendingLogin` is set. You don't dismiss it.
2. You switch to the workspace B and run a request that triggers its own login modal. Now the request would wait for the modal from the workspace A.

But should we assume that the user can change a workspace when the login modal is visible? I think it's not a realistic scenario.

Reply:
I think that makes sense.

This made me realize that if a relogin modal is shown and you open a deep link to another workspace and that deep link triggers a login modal, then the relogin attempt will just hang forever.

I guess now that we have this `pendingLogin`, maybe we could actually make it so that `openRegularDialog`/`openImportantDialog` cancels the previous dialog when opening a new one?

But also feel free to ignore it as it goes a little bit out of scope of this PR, which was made to solve the issue at hand with concurrent requests in unified resources.

Reply:
Good idea, it shouldn't have any negative side effects. I clicked through the app and everything seems to work fine.

Reply:
I did some tests too and everything appears to be working fine. 👍